File: todo-api-rs/src/application/functions/todo/create.rs
use crate::application::dtos::todo::create::{CreateTodoError, CreateTodoInput};
use crate::application::repositories::todo::TodoRepository;
use crate::domain::entities::todo::TodoEntity;
use crate::domain::types::{DateTime, Id};

pub async fn create_todo<T>(
    ctx: CreateTodoContext<'_, T>,
    input: CreateTodoInput,
) -> Result<TodoEntity, CreateTodoError>
where
    T: TodoRepository,
{
    let current_dt = DateTime::new();
    let todo_entity = TodoEntity {
        id: Id::new(),
        title: input.title,
        description: input.description,
        status: input.status,
        todo_at: input.todo_at,
        tags: Vec::new(),
        created_at: current_dt,
        updated_at: current_dt,
    };

    ctx.todo_repository
        .create(todo_entity)
        .await
        .map_err(CreateTodoError::Creating)
}

#[derive(Clone, Debug)]
pub struct CreateTodoContext<'a, T>
where
    T: TodoRepository,
{
    pub todo_repository: &'a T,
}

File: todo-api-js/src/application/use-cases/todo/create.ts
import { InternalError } from "@domain/utils/exception";
import { Err } from "@domain/utils/result";
import { TodoEntity } from "@domain/entities/todo";
import { type UseCase } from "@domain/use-case";
import { type CreateTodoInput, type CreateTodoOutput } from "@application/dtos/todo/create";
import { type TodoRepository } from "@application/repositories/todo";

export class CreateTodoUseCase implements UseCase<CreateTodoInput, CreateTodoOutput> {
    constructor(private repository: TodoRepository) {}

    async exec(input: CreateTodoInput): Promise<CreateTodoOutput> {
        try {
            const entity = TodoEntity.create({
                title: input.title,
                description: input.description,
                status: input.status,
                todoAt: input.todoAt,
            });

            const result = await this.repository.create(entity);
            return result.map(() => entity);
        } catch (e) {
            const error = new InternalError(e instanceof Error ? e : undefined);
            return new Err(error);
        }
    }
}
